### PR TITLE
SPV: only declare the pure 8/16-bit capabilities when needed.

### PR DESCRIPTION
**Pull request by johnkslang**
Only when operations stray outside the 8/16-bit storage-capabilities
are the general (pure 8/16-bit) capabilities needed.

### CODE REVIEW

File: SPIRV/SpvPostProcess.cpp [hunk 1]
```diff
@@ -61,8 +61,82 @@ namespace spv {
...
+            }
+        }
+        break;
+    case OpAccessChain:
+    case OpPtrAccessChain:
+    case OpCopyObject:
+    case OpConvertFToU:
+    case OpConvertFToS:
+    case OpConvertSToF:
+    case OpConvertUToF:
```
Comment (jeffbolznv) [inline SPIRV/SpvPostProcess.cpp]: I don't think FTo{U,S} or {S,U}ToF are supported, just {F,S,U}Convert.

Comment (johnkslang) [reply SPIRV/SpvPostProcess.cpp]: Right, I discovered this testing. I think the extension is wrong. I uncovered quite a few problems with the extension, which I go into deeper in GitLab.

For this one, I think it is perfectly reasonable to do the following, and it generates one of the additional instructions I subtracted:

```
#version 460
#extension GL_KHX_shader_explicit_arithmetic_types_int8: require

struct S1 {
    bool b;
    int8_t i8;
};

struct S2 {
    int i;
    S1 s1;
};

buffer ub {
    S2 s2;
} buff;

void main()
{
    float f = buff.s2.s1.i8;
}
```

Comment (johnkslang) [reply SPIRV/SpvPostProcess.cpp]: My example is meant as pseudo SPIR-V, don't get hung up on the #extension, etc.

The intention is that a SPIR-V shader that does that only needs the int8-storage capability, not the int8 capability.

If it was the only added instruction, maybe I shouldn't do it, but there were many instructions I needed to add to make things actually work.

Comment (jeffbolznv) [reply SPIRV/SpvPostProcess.cpp]: For better or worse, the SPV extension has always been clear that the width-only conversions (F,S,U) are supported and the others are not. I think it's much too late to change that. GL_EXT_shader_16bit_storage avoids needing these instructions by forbidding implicit conversions and limiting the constructors that are supported for these types. So I don't think there is any actual demand for them coming from GLSL.

Comment (johnkslang) [reply SPIRV/SpvPostProcess.cpp]: Okay, I can live with that.

I think OpAccessChain will be another story.

Comment (jeffbolznv) [reply SPIRV/SpvPostProcess.cpp]: I agree that the spec is unclear/misleading about access chains' usage of these pointer types (and other things...). But there is CTS test coverage that uses them, so it is OK to clarify the spec to explicitly allow those.

Comment (johnkslang) [reply SPIRV/SpvPostProcess.cpp]: Extra conversions are now removed.

**Pull request merged**